pkg/groupsnapshot/controllers/groupsnapshot.go

```diff
@@ -0,0 +1,615 @@
+	var err error
+
+	// Pre checks
+	if len(groupSnap.Spec.PVCSelector.MatchExpressions) > 0 {
+		err = fmt.Errorf("matchExpressions are currently not supported in the spec. Use matchLabels")
+	}
+
+	if len(groupSnap.Spec.PVCSelector.MatchLabels) == 0 {
+		err = fmt.Errorf("matchLabels are required for group snapshots. Refer to spec examples")
+	}
+
+	if err != nil {
+		groupSnap.Status.Status = stork_api.GroupSnapshotFailed
+		groupSnap.Status.Stage = stork_api.GroupSnapshotStageFinal
+		return updateCRD, err
+	}
+
+	_, err = k8sutils.GetPVCsForGroupSnapshot(groupSnap.Namespace, groupSnap.Spec.PVCSelector.MatchLabels)
+	if err != nil {
+		if groupSnap.Status.Status == stork_api.GroupSnapshotPending {
+			return !updateCRD, err
+		}
+
+		groupSnap.Status.Status = stork_api.GroupSnapshotPending
+		groupSnap.Status.Stage = stork_api.GroupSnapshotStagePrechecks
+	} else {
+		groupSnap.Status.Status = stork_api.GroupSnapshotInProgress
+		// done with pre-checks, move to pre-snapshot stage
+		groupSnap.Status.Stage = stork_api.GroupSnapshotStagePreSnapshot
+	}
+
+	return updateCRD, err
+}
+
+func (m *GroupSnapshotController) handlePreSnap(groupSnap *stork_api.GroupVolumeSnapshot) (bool, error) {
+	ruleName := groupSnap.Spec.PreSnapshotRule
+	if len(ruleName) == 0 {
+		groupSnap.Status.Status = stork_api.GroupSnapshotInProgress
+		// No rule, move to snapshot stage
+		groupSnap.Status.Stage = stork_api.GroupSnapshotStageSnapshot
+		return updateCRD, nil
+	}
+
+	log.GroupSnapshotLog(groupSnap).Infof("Running pre-snapshot rule: %s", ruleName)
+	r, err := k8s.Instance().GetRule(ruleName, groupSnap.Namespace)
+	if err != nil {
+		return !updateCRD, err
+	}
+
+	pvcs, err := k8sutils.GetPVCsForGroupSnapshot(groupSnap.Namespace, groupSnap.Spec.PVCSelector.MatchLabels)
+	if err != nil {
+		return !updateCRD, err
+	}
+
+	backgroundCommandTermChan, err := rule.ExecuteRule(r, rule.PreExecRule, groupSnap, pvcs)
+	if err != nil {
+		if backgroundCommandTermChan != nil {
+			backgroundCommandTermChan <- true // terminate background commands if running
+		}
+
+		return !updateCRD, err
+	}
+
+	if backgroundCommandTermChan != nil {
+		snapUID := string(groupSnap.ObjectMeta.UID)
+		m.bgChannelsForRules[snapUID] = backgroundCommandTermChan
+	}
+
+	// done with pre-snapshot, move to snapshot stage
+	groupSnap.Status.Stage = stork_api.GroupSnapshotStageSnapshot
+	return updateCRD, nil
+}
+
+func (m *GroupSnapshotController) handleSnap(groupSnap *stork_api.GroupVolumeSnapshot) (bool, error) {
+	var (
+		err      error
+		stage    stork_api.GroupVolumeSnapshotStageType
+		status   stork_api.GroupVolumeSnapshotStatusType
+		response *volume.GroupSnapshotCreateResponse
+	)
+
+	if len(groupSnap.Status.VolumeSnapshots) > 0 {
+		logrus.Infof("Group snapshot already active. Checking status")
+		response, err = m.Driver.GetGroupSnapshotStatus(groupSnap)
+		if err != nil {
+			logrus.Errorf("group snapshot status returned err: %v", err)
+			m.Recorder.Event(groupSnap,
```

Comment:
This will log multiple events? Handle() is also recording events when this function returns an error.

Same for other places in this function where event is being recorded.